### PR TITLE
application: implement permissions

### PR DESCRIPTION
**Pull request by Garfield-fr**
* Removes the role mechanism.
* Implements the permission mechamism.
* Adds some tests.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## Dependencies

- https://github.com/zannkukai/rero-ils/tree/zan-expose-permissions

### CODE REVIEW

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html [hunk 1]
```diff
@@ -253,6 +253,7 @@ <h3 id="{{ 'doc-altgr-title-' + i }}">{{ altgr_title }}</h3>
         *ngIf="!record.metadata.harvested && record.metadata.pid"
         id="documents-local-field-tab"
         tabOrder="4"
+        [customClass]="showhideLocalFieldsTab"
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html:256]: why not use this condition into the `*ngIf` ?
Additionally the name of the function could be more elegant and return a boolean value

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.ts [hunk 2]
```diff
@@ -131,6 +132,19 @@ export class DocumentDetailViewComponent implements DetailRecord, OnInit, OnDest
...
+  }
+
+  /**
+   * Show or hide local fields tab
+   * @return string - hide with class 'd-none'
+   */
+  get showhideLocalFieldsTab(): string {
+    return this._permissionsService.canAccess([PERMISSIONS.LOFI_SEARCH, PERMISSIONS.LOFI_CREATE]) ? '': 'd-none';
+  }
+
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.ts:147]: See remarks above (into the related template file)

File: projects/admin/src/app/record/detail-view/document-detail-view/holdings/holdings.component.ts [hunk 3]
```diff
@@ -84,6 +84,11 @@ export class HoldingsComponent implements OnInit {
     return this._userService.user.currentOrganisation;
   }
 
+  /** return all available permissions for current user */
+  get permissions(): IPermissions {
+    return PERMISSIONS;
+  }
+
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/document-detail-view/holdings/holdings.component.ts:91]: same

File: projects/admin/src/app/record/detail-view/holding-detail-view/serial-holding-detail-view/serial-holding-detail-view.component.html [hunk 4]
```diff
@@ -71,19 +72,21 @@ <h1 class="mb-2" translate>Holdings</h1>
           <ng-container actions *ngIf="last">
             <button type="button" class="btn btn-sm btn-outline-primary ml-1"
                     title="{{ 'Quick receive' | translate }}"
+                    [permissions]="permissions.ITEM_CREATE"
                     (click)="quickIssueReceive()">
-              <i class="fa fa-file-text-o"></i>
+              <i class="fa fa-file-text"></i>
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/holding-detail-view/serial-holding-detail-view/serial-holding-detail-view.component.html:77]: I think this icon was especially choose by PO, same for `fa-pencil` below (you replaced by `fa-plus`)

File: projects/admin/src/app/record/detail-view/holding-detail-view/serial-holding-detail-view/serial-holding-detail-view.component.html [hunk 5]
```diff
@@ -122,7 +125,7 @@ <h1 class="mb-2" translate>Holdings</h1>
   <!-- END OF DETAILS TAB ---------------------------------------------------->
 
   <!-- LOCAL FIELDS TAB ------------------------------------------------------------>
-  <tab id="holding-local-field-tab" tabOrder="3">
+  <tab id="holding-local-field-tab" tabOrder="3" [customClass]="showhideLocalFieldsTab">
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/holding-detail-view/serial-holding-detail-view/serial-holding-detail-view.component.html:128]: `*ngIf` or your `[permissions]` directive (seems better)?

File: projects/admin/src/app/record/detail-view/holding-detail-view/serial-holding-detail-view/serial-holding-detail-view.component.ts [hunk 6]
```diff
@@ -61,6 +61,19 @@ export class SerialHoldingDetailViewComponent implements OnInit {
...
+   */
+  get showhideLocalFieldsTab(): string {
+    return this._permissionsService.canAccess([PERMISSIONS.LOFI_SEARCH, PERMISSIONS.LOFI_CREATE]) ? '': 'd-none';
+  }
+
+  /** return all permissions */
+  get permissions(): IPermissions {
+    return PERMISSIONS;
+  }
+
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/holding-detail-view/serial-holding-detail-view/serial-holding-detail-view.component.ts:76]: same + same = same^2

File: projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts [hunk 7]
```diff
@@ -81,6 +81,14 @@ export class ItemDetailViewComponent implements DetailRecord, OnInit, OnDestroy
...
   }
 
+  /**
+   * Permissions
+   * @return Object with all permissions
+   */
+  get permissions(): IPermissions {
+    return PERMISSIONS;
+  }
+
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts:91]: same

File: projects/admin/src/app/record/detail-view/item-detail-view/item-transactions/item-transactions.component.ts [hunk 8]
```diff
@@ -42,6 +42,11 @@ export class ItemTransactionsComponent implements OnInit {
   /** Requested loan(s) */
   requestedBy: Array<any> = [];
 
+  /** return all permissions */
+  get permissions(): IPermissions {
+    return PERMISSIONS;
+  }
+
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/item-detail-view/item-transactions/item-transactions.component.ts:49]: same

File: projects/admin/src/app/record/detail-view/local-field/local-field.component.ts [hunk 9]
```diff
@@ -43,6 +43,11 @@ export class LocalFieldComponent implements OnInit {
     items: 'item'
   };
 
+    /** return all available permissions for current user */
+    get permissions(): IPermissions {
+      return PERMISSIONS;
+    }
+
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/local-field/local-field.component.ts:50]: same

File: projects/admin/src/app/record/detail-view/patron-detail-view/patron-detail-view.component.ts [hunk 10]
```diff
@@ -85,19 +85,16 @@ export class PatronDetailViewComponent implements OnInit, DetailRecord, OnDestro
     return data.sort((a, b) => b.weight - a.weight);
   }
 
-  /** Check if the patron has the librarian role */
-  get isLibrarian(): boolean {
-    return this.patron && this.patron.roles && this.patron.roles.includes('librarian');
+  get permissions(): IPermissions {
+    return PERMISSIONS;
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/patron-detail-view/patron-detail-view.component.ts:89]: same or missing docstring

File: projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.html [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+  You should have received a copy of the GNU Affero General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+<div class="container" *ngIf="permissionsByRoles">
+  <h1 translate>Permissions matrix</h1>
+    <div class="my-3">
+      <label for="filter" translate>Filter permission list</label>
+      <input id="filter" class="form-control" #search (keyup)="filterPermissions(search.value)" />
+      <small id="filterInfo" class="form-text text-muted">Enter a part of the permission name to filter the list.</small>
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.html:23]: `id` attribute could be removed

File: projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.html [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+          </tr>
+        </thead>
+        <tbody>
+            <tr *ngFor="let permission of filteredPermissions">
+              <td class="permission">{{ permission }}</td>
+              <td *ngFor="let role of roles" class="access text-center">
+                <span title="{{ role | translate }}">
+                  <i
+                    class="fa fa-square"
+                    [ngClass]="{'text-success': permissionsByRoles[role][permission], 'text-danger': !permissionsByRoles[role][permission]}">
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.html:42]: As mentionned this morning, a third value is possible.
In such case (more than 2 choices), it's more readable if each class is on a separate line (my opinion)

File: projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.scss [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  max-height: 200px;
+}
+
+.permission {
+  width: 200px;
+}
+
+.access {
+  width: 30px;
+}
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.scss:48]: move this into the `table` section (as `th`)

File: projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
+/*
```
Comment (zannkukai) [inline projects/admin/src/app/record/detail-view/permission-detail-view/permission-detail-view.component.spec.ts:1]: Nothing is tested, could be removed

File: projects/admin/src/app/record/operation-logs/operation-logs-dialog/operation-logs-dialog.component.ts [hunk 15]
```diff
@@ -33,6 +34,11 @@ export class OperationLogsDialogComponent {
   /** Modal ref */
   bsModalRef: BsModalRef;
 
+  /** return all permissions */
+  get permissions(): IPermissions {
+    return PERMISSIONS;
+  }
+
```
Comment (zannkukai) [inline projects/admin/src/app/record/operation-logs/operation-logs-dialog/operation-logs-dialog.component.ts:41]: same (next I wrote a joke to not be too redundant)

File: projects/admin/src/app/service/app-initializer.service.ts [hunk 16]
```diff
@@ -62,7 +62,7 @@ export class AppInitializerService {
     return this._userService.load().pipe(
       tap((user: User) => {
         this._typeaheadFactoryService.init();
-        if (user.isAuthorizedAdminAccess) {
+        if (user.isAdminUiAccess) {
```
Comment (zannkukai) [inline projects/admin/src/app/service/app-initializer.service.ts:65]: hasAdminUIAccess as previously suggested

File: projects/shared/src/lib/class/user.ts [hunk 17]
```diff
@@ -89,11 +88,11 @@ export class User implements IUser {
   }
 
   /**
-   * Is authorized admin access
+   * Is granted to access to admin UI
    * @return boolean
    */
-  get isAuthorizedAdminAccess(): boolean {
-    return this._patronLibrarian !== undefined;
+  get isAdminUiAccess(): boolean {
```
Comment (zannkukai) [inline projects/shared/src/lib/class/user.ts:94]: `hasAdminUIAccess` ?

File: projects/shared/src/lib/class/user.ts [hunk 18]
```diff
@@ -117,23 +116,7 @@ export class User implements IUser {
...
-    return this._patronRoles.includes('librarian');
-  }
-
-  /**
-   * Is the user a system_librarian
-   * @return boolean
-   */
-  get isSystemLibrarian(): boolean {
-    return this._patronRoles.includes('system_librarian');
+    return this.patrons?.some((patron: IPatron) => 'patron' in patron);
```
Comment (zannkukai) [inline projects/shared/src/lib/class/user.ts:119]: don't you have an enum class to list all PatronRole ? Should be better than use 'patron' string here

File: projects/shared/src/lib/directive/link-permissions.directive.ts [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  // eslint-disable-next-line @angular-eslint/directive-selector
+  selector: '[linkPermissions]',
+})
+export class LinkPermissionsDirective implements AfterViewInit {
+
+  /** permissions */
+  private perms: string[] = [];
+
+  /** operator */
+  private andOperator: boolean = false;
```
Comment (zannkukai) [inline projects/shared/src/lib/directive/link-permissions.directive.ts:30]: always the same ;-) And now the joke (in French) :
 
"Tu sais pourquoi Mickey Mousse ? parce que Mario Brosse !"

File: projects/shared/src/lib/service/permissions.service.ts [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+   * @returns this
+   */
+  setPermissions(permissions: string[]): this {
+    permissions.map((permission: string) => {
+      if (!this.permissions.has(permission)) {
+        this.permissions.add(permission);
+      }
+    });
+
+    return this;
```
Comment (zannkukai) [inline projects/shared/src/lib/service/permissions.service.ts:41]: remove blank line.
Are you sure about return statement ? maybe `return this._permissions` ?

File: projects/shared/src/lib/service/permissions.service.ts [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+import { Injectable } from '@angular/core';
+import { PERMISSIONS } from '../util/permissions';
+
+@Injectable({
+  providedIn: 'any'
+})
+export class PermissionsService {
+
+  /** Permissions set */
+  private permissions = new Set<string>();
```
Comment (zannkukai) [inline projects/shared/src/lib/service/permissions.service.ts:27]: add "_" for private variable

File: projects/shared/src/lib/service/permissions.service.ts [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+    return this;
+  }
+
+  /**
+   * Can access
+   * @param permission - permission or array of permissions
+   * @param and - operator
+   * @returns true if permission is granted
+   */
+   canAccess(permission: string | string[], and: boolean = false): boolean {
```
Comment (zannkukai) [inline projects/shared/src/lib/service/permissions.service.ts:50]: It's here the core of MY "andOperator" problem. If you change `and: boolean` by `operator: string = 'or'`, you could much more flexible if necessary in next PR's/US.

File: projects/shared/src/lib/service/permissions.service.ts [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+    if (typeof permission === 'string') {
+      permission = [permission];
+    }
+
+    return and
+      ? permission.every((perm: string) => this.permissions.has(perm))
+      : permission.some((perm: string) => this.permissions.has(perm))
```
Comment (zannkukai) [inline projects/shared/src/lib/service/permissions.service.ts:57]: use a switch to implement as many operator as you want
```
swtich (operator) {
    case 'and': 
        return permission.every((perm: string) => this.permissions.has(perm))
    case 'or': 
        return permission.some((perm: string) => this.permissions.has(perm))
    case 'xor':
        ... # should not be implemented at this time ;-)
    default:
        throw new Exception("invalid operator")

File: projects/shared/src/lib/service/user.service.ts [hunk 24]
```diff
@@ -68,7 +72,12 @@ export class UserService {
         const settings: ISettings = loggedUser.settings;
         this._appSettingsService.settings = settings;
         delete loggedUser[settingsKey];
-        this._user = new User(loggedUser, this._appSettingsService.librarianRoles);
+        if (loggedUser.permissions) {
+          delete loggedUser.permissions[loggedUser.permissions.indexOf(PERMISSIONS.CIRC_ADMIN)]
+          loggedUser.permissions.push(PERMISSIONS.PTRN_ACCESS);
```
Comment (zannkukai) [inline projects/shared/src/lib/service/user.service.ts:77]: Seems use for TEST & debug, Must be removed

**Review by zannkukai — CHANGES_REQUESTED** [review]
Sorry for all my remarks. But it's a huge and 👏  thx

File: projects/admin/src/app/circulation/circulation-routing.module.ts [hunk 25]
```diff
@@ -44,40 +46,51 @@ const routes: Routes = [
       },
       {
         path: 'loan',
-        component: LoanComponent
+        component: LoanComponent,
+        canActivate: [ PermissionGuard ], data: { permissions: [ PERMISSIONS.CIRC_ADMIN ] }
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/circulation/circulation-routing.module.ts:50]: When there are children, the parent route is never used

File: projects/admin/src/app/menu/service/menu-user-services.service.ts [hunk 26]
```diff
@@ -160,27 +179,31 @@ export class MenuUserServicesService extends MenuBase {
...
+      const createBibRecordMenu = catalogMenu.addChild(
+        this._translateService.instant('Create a bibliographic record')
+      )
+      .setRouterLink(['/', 'records', 'documents', 'new'])
+      .setAttribute('id', 'create-bibliographic-record-menu')
+      .setExtra('iconClass', 'fa fa-file-o');
+      this._translatedName(createBibRecordMenu, 'Create a bibliographic record');
+
+      // ----- IMPORT FROM WEB
+      const importFromWebMenu = catalogMenu.addChild('Import from the web')
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/service/menu-user-services.service.ts:201]: Changed with `doc-create`

File: projects/admin/src/app/menu/service/menu-user-services.service.ts [hunk 27]
```diff
@@ -202,160 +225,234 @@ export class MenuUserServicesService extends MenuBase {
...
+    if (this._permissionsService.canAccess([
+      PERMISSIONS.ITEM_ACCESS
+    ])) {
+      const reportMenu = menu.addChild('Reports & monitoring')
+      .setAttribute('id', 'report-monitoring-menu')
+      .setExtra('iconClass', 'fa fa-bar-chart');
+      this._translatedName(reportMenu, 'Reports & monitoring');
+
+      // ----- INVENTORY LIST
+      if (this._permissionsService.canAccess(PERMISSIONS.ITEM_ACCESS)) {
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/service/menu-user-services.service.ts:305]: It is in the planning to put other menu lines.

**Review by zannkukai — APPROVED** [review]
There is yet many `if` statement into `[menu-user-services.service.ts](https://github.com/rero/rero-ils-ui/pull/904/files#diff-1f81d18441a4bf8ce4ced540eb343d9464cbe66f2c5681b657c44663a15a0f99)`. Are they removed in next PR (including the recursing function ?)

**Pull request merged**